### ISSUE TITLE
bit extraction breaks with -mx32 -march=core-avx2

### ISSUE THREAD

**Issue by tycho** [Cyan4973/zstd#178]
__builtin_ia32_bextr_u64 seems to break builds with the X32 ABI:

fullbench32 built with -mx32 -march=core-avx2:
```
./fullbench32
*** Zstandard speed analyzer 0.6.0 32-bits, by Yann Collet (Apr 28 2016) ***
 Sample 10MiB :                                                                
 1- ZSTD_compress                  :   330.8 MB/s  (  3174336)
ERROR ! ZSTD_decompress() => Corrupted block detected !!
```
If I make this change, then fullbench32 runs to completion without issues:
```diff
diff --git a/lib/bitstream.h b/lib/bitstream.h
index fd114e5..357706d 100644
--- a/lib/bitstream.h
+++ b/lib/bitstream.h
@@ -53,14 +53,6 @@ extern "C" {
 #include "error_private.h"  /* error codes and messages */
 
 
-/*=========================================
-*  Target specific
-=========================================*/
-#if defined(__BMI__) && defined(__GNUC__)
-#  include <immintrin.h>   /* support for bextr (experimental) */
-#endif
-
-
 /*-******************************************
 *  bitStream encoding API (write forward)
 ********************************************/
@@ -300,11 +292,7 @@ MEM_STATIC size_t BIT_getUpperBits(size_t bitD, U32 const start)
 
 MEM_STATIC size_t BIT_getMiddleBits(size_t bitD, U32 const nbBits, U32 const start)
 {
-#if defined(__BMI__) && defined(__GNUC__)   /* experimental */
-    return __builtin_ia32_bextr_u64(bitD, (nbBits<<8) | start );
-#else
     return (bitD >> start) & BIT_mask[nbBits];
-#endif
 }
 
 MEM_STATIC size_t BIT_getLowerBits(size_t bitD, U32 const nbBits)
@@ -321,12 +309,8 @@ MEM_STATIC size_t BIT_getLowerBits(size_t bitD, U32 const nbBits)
  */
  MEM_STATIC size_t BIT_lookBits(const BIT_DStream_t* bitD, U32 nbBits)
 {
-#if defined(__BMI__) && defined(__GNUC__)   /* experimental */
-    return __builtin_ia32_bextr_u64(bitD->bitContainer, (nbBits<<8) | (64 - bitD->bitsConsumed - nbBits) );
-#else
     U32 const bitMask = sizeof(bitD->bitContainer)*8 - 1;
     return ((bitD->bitContainer << (bitD->bitsConsumed & bitMask)) >> 1) >> ((bitMask-nbBits) & bitMask);
-#endif
 }
 
 /*! BIT_lookBitsFast() :
```

The __builtin_ia32_bextr_u64 extraction crashes the compiler with -m32 -march=core-avx2:

https://gcc.gnu.org/bugzilla/show_bug.cgi?id=70858

Apparently you shouldn't be using __builtin_ia32_bextr_u64 and should instead be using the _bextr_u64 intrinsic. And the 'u64' variant is unavailable on -m32 builds, so you should be using _bextr_u32. But if you use _bextr_u32, then you always get that "corrupted block detected" failure (on -m32, -mx32, *or* -m64). So that's fun...

I also noticed that the bitstream API incorrectly (through use of size_t) believes the native register type is 32-bits wide on -mx32 (in actuality, the full 64-bit register set is available). I'd recommend doing something like this, and then fixing all the places that assume sizeof(size_t) == sizeof(native register size)

```c
#ifdef __x86_64__
/* On LP64/ILP32, use 64-bit registers. */
typedef U64 nint_t;
#else
/* On everything else, assume that sizeof(size_t) == the native register size. */
typedef size_t nint_t;
#endif
```
(where "nint_t" stands for "native integer type")

**Comment by tycho** [comment]
Ah, I missed the constant '64' in BIT_lookBits, that'd explain the breakage on -m32 with _bextr_u32. This makes -m32, -mx32, and -m64 work correctly:
```diff
diff --git a/lib/bitstream.h b/lib/bitstream.h
index fd114e5..f955b87 100644
--- a/lib/bitstream.h
+++ b/lib/bitstream.h
@@ -301,7 +301,11 @@ MEM_STATIC size_t BIT_getUpperBits(size_t bitD, U32 const start)
 MEM_STATIC size_t BIT_getMiddleBits(size_t bitD, U32 const nbBits, U32 const start)
 {
 #if defined(__BMI__) && defined(__GNUC__)   /* experimental */
-    return __builtin_ia32_bextr_u64(bitD, (nbBits<<8) | start );
+#if defined(_x86_64__) && !defined(__ILP32__)
+    return _bextr_u64(bitD, start, nbBits);
+#else
+    return _bextr_u32(bitD, start, nbBits);
+#endif
 #else
     return (bitD >> start) & BIT_mask[nbBits];
 #endif
@@ -322,7 +326,11 @@ MEM_STATIC size_t BIT_getLowerBits(size_t bitD, U32 const nbBits)
  MEM_STATIC size_t BIT_lookBits(const BIT_DStream_t* bitD, U32 nbBits)
 {
 #if defined(__BMI__) && defined(__GNUC__)   /* experimental */
-    return __builtin_ia32_bextr_u64(bitD->bitContainer, (nbBits<<8) | (64 - bitD->bitsConsumed - nbBits) );
+#if defined(__x86_64__) && !defined(__ILP32__)
+    return _bextr_u64(bitD->bitContainer, (sizeof(bitD->bitContainer) * 8) - bitD->bitsConsumed - nbBits, nbBits);
+#else
+    return _bextr_u32(bitD->bitContainer, (sizeof(bitD->bitContainer) * 8) - bitD->bitsConsumed - nbBits, nbBits);
+#endif
 #else
     U32 const bitMask = sizeof(bitD->bitContainer)*8 - 1;
     return ((bitD->bitContainer << (bitD->bitsConsumed & bitMask)) >> 1) >> ((bitMask-nbBits) & bitMask);
```

Still, -mx32 builds underperform due to the sizeof(size_t) on that target.

**Comment by Cyan4973** [comment]
Thanks for notification @tycho .

Indeed, this experimental mode was added without much testing, and I made the naive assumption that BMI support implied 64-bits mode. Never tested on 32-bits mode, and as you found, it was an error, as this combination is nonetheless possible and buggy in this case.
Thanks for clear guidance. I will look into it.

The second point would likely deserve another entry,
It's very interesting : separating native register size from `size_t`, which were assumed to be the same so far. It turns out to be wrong too, and would likely deserve a shake up.
For this point, it would also be best to understand in which cases this assumption is wrong, for better testings. I'm not sure yet to understand the full implications of `-mx32` and `__ILP32__` modes.

**Comment by tycho** [comment]
Hi @Cyan4973!

Sadly the x32 ABI is not widely used, but it's very nice for limiting memory utilization on 64-bit systems -- it basically provides programs with access to the full 64-bit register set, but uses 32-bit pointers. Wikipedia has a [handy table](https://en.wikipedia.org/wiki/X32_ABI#Details) that shows the differences between `-m32`, `-mx32`, and `-m64`. This has a nice impact on binary size, runtime memory utilization, and occasionally performance. It does make some sense that `size_t` would be 32-bit on x32, since the address space is 4GB (`UINT32_MAX`).

I put up some commits in my zstd fork which make things work nicely on all 3 ABIs: https://github.com/Cyan4973/zstd/compare/master...tycho:master

It would also be nice to get the `-m32` builds to match performance of the `-m64` builds, but that's probably a 3rd issue to track. I'm puzzled as to why GCC doesn't choose to use XMM registers for 64-bit integer arithmetic on 32-bit builds with `-msse2` enabled. Would likely be cheaper than `add`+`adc` and such. I know the Intel compiler does that in some cases, as can be seen here:

GCC 6.1: https://tinyurl.com/ho3amsl
ICC 16: https://tinyurl.com/hbgovdd

And that's a trick that can be done as far back as `-march=pentium4`, since SSE2 was introduced so many processor generations ago.

**Comment by Cyan4973** [comment]
Correct behavior of BMI instruction on `x32` and `x86` target seems to be properly fixed with 6f9c056662cfd29a44dbea3bfcaf38c2dffe5d9e .

Unfortunately `_bextr_u64()` and `_bextr_u32()` do not seem supported on the [gcc version of Travis CI](https://travis-ci.org/Cyan4973/zstd/jobs/126995255), which is a pity as it will hamper continuous test ability for BMI code branch.

No change yet on register type, still lazily evaluated as `size_t`, which is sub-optimal for `x32`. But the impact on the code is non negligible, so I want to keep that task for later on.

**Comment by tycho** [comment]
It surprises me that their GCC version doesn't support the `_bextr_u32` or `_bextr_u64` intrinsics:

```
$ grep -Rl _bextr_u64  /usr/lib/gcc/x86_64-unknown-linux-gnu/*/include
/usr/lib/gcc/x86_64-unknown-linux-gnu/4.6.4/include/bmiintrin.h
/usr/lib/gcc/x86_64-unknown-linux-gnu/4.7.4/include/bmiintrin.h
/usr/lib/gcc/x86_64-unknown-linux-gnu/4.8.5/include/bmiintrin.h
/usr/lib/gcc/x86_64-unknown-linux-gnu/4.9.3/include/bmiintrin.h
/usr/lib/gcc/x86_64-unknown-linux-gnu/5.3/include/bmiintrin.h
/usr/lib/gcc/x86_64-unknown-linux-gnu/6.1.0/include/bmiintrin.h
```

What ancient GCC version are they running?

**Comment by Cyan4973** [comment]
According to the log : 
```
gcc --version
gcc (Ubuntu/Linaro 4.6.3-1ubuntu5) 4.6.3
Copyright (C) 2011 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by tycho** [comment]
Weird. I would expect it to work. Try including x86intrin.h instead of immintrin.h?

**Comment by tycho** [comment]
FYI moved the x32 patches to my 'dev' branch and rebased them on your current 'dev' branch: https://github.com/Cyan4973/zstd/compare/dev...tycho:dev

**Comment by Cyan4973** [comment]
BMI issue fixed in v0.6.1

For the other pb of 64-bits performance under x32,
maybe we need to open another topic
